docs/technical-doc/storage.md

```diff
@@ -0,0 +1,167 @@
+# Object storage management
+
+## The `massa-storage` crate and the `Storage` structure it provides
+
+### General description
+
+This crate provides the `Storage` structure which give access to a globally shared object store containing blocks, endorsements and operations.
+It also allows locally claiming references to a list of shared objects for a particular instance of `Storage`.
+When no instance of `Storage` owns a reference to a given object anymore, that object is automatically dropped from the globally shared object store. 
+
+The `Storage` structure contains:
+* shared references to global maps containing objects like blocks, endorsements, operations and indices allowing to retrieve them by various criteria
+* shared reference counters to those objects
+* a local (non-shared) list of object references owned by that particular instance of `Storage`
+
+### Contructing and cloning `Storage`
+
+An initial `Storage` object should be created in the program's entry point using:
+
+```rust
+let storage: Storage = Default::default();
+```
+
+**All subsequent instances of `Storage` used throughout the program should always be clones of that instance.**
+
+There are two ways of cloning an instance of `Storage`:
+* `storage.clone()` will clone an existing instance and claim all the local references to objects that the initial instance held
+* `storage.clone_without_refs()` will clone an existing instance but the resulting instance will not hold any local references to objects
+
+### Adding an element to the global shared object store
+
+Here we take the example of adding a set of operations to the global shared object store.
+Note that the same applies to blocks and endorsements.
+
+```rust
+let operations: Vec<WrappedOperation> = XXXX;
+storage.store_operations(operations);
+// Here, all `operations` were stored inside the global shared object store and `storage` has acquired local references to all of them. Objects that were already stored beforehand are not overwritten, but a local reference is simply acquired. 
+```
+
+### Claiming a local reference to an object that is already in the globally shared store
+
+Here we take the example of claiming a local reference to a set of operations that is already in the global shared object store.
```

Comment:
I think `local reference` could be renamed simple "reference", because it is in fact a global reference count, so it's not really "local", it's just that one instance of storage claiming a reference it seems. "Local" makes it sound like it would have no effect globally.

Reply:
renamed